ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/snmp-settings.j2

```diff
@@ -102,18 +117,27 @@ snmp-server location {{ snmp_server.location }}
 {%             if user.group is arista.avd.defined %}
 {%                 set user_cli = user_cli ~ " " ~ user.group %}
 {%             endif %}
+{%             if user.remote_ip is arista.avd.defined %}
+{%                 set user_cli = user_cli ~ " remote " ~ user.remote_ip %}
+{%                 if user.udp_port is arista.avd.defined %}
+{%                     set user_cli = user_cli ~ " udp-port " ~ user.udp_port %}
+{%                 endif %}
+{%             endif %}
 {%             if user.version is arista.avd.defined %}
 {%                 set user_cli = user_cli ~ " " ~ user.version %}
 {%             endif %}
 {%             if user.auth is arista.avd.defined
                   and user.version is arista.avd.defined('v3')
                   and user.auth_passphrase is arista.avd.defined %}
```

Comment:
Maybe we should rethink this condition (the 2nd and which checks if passphrase is set). The CLI allows the following:

just setting "v3" without any auth or priv pass

it also allows just setting "v3 and localized BLABLA" (it will accept this, but omit the localized in the config)

=======

To me it would make sense to build the CLI step by step, so "plain" v3 only would be allowed, "v3 localized xxx" would be allowed, just auth allowed (no priv specified) and the same with localized....

most easiest would be to simply construct the cli from whatever the user inputs.. and keep the conditions minimal.

What do you think?

Reply:
Will the v3 alone work?
We already discussed the localized option alone, but since it is removed by EOS, we decided to to keep that.

Reply:
yup, just v3 works.. also appears like that in the running config:

```
snmp-server engineID local 617269737461
snmp-server group bla v3 priv
snmp-server user bla test v3
```

you can use unauthenticated snmpv3. In general you just need a username to use snmpv3.. kinda defeats the purpose of it, but it is possible.

What troubles me more (and might be worth addressing to EOS developers) is that EOS currently accepts auth passwords < 8 characters (cleartext). 

Problem occurs if you want to query these agents with netsnmp for example:

```
Error: passphrase chosen is below the length requirements of the USM (min=8).
snmpwalk:  (The supplied password length is too short.)
Error generating a key (Ku) from the supplied authentication pass phrase.
```

At least SHA seems to require at least 8 character password. Might make sense to enforce 8char min on OS level, or?

Reply:
with snmpv3 they tried to make up for everything that snmpv2 was missing while also still allowing what snmpv2 allowed.. (my opinion).

thats why you have: unauthenticated, authenicated but no priv, and authpriv and so on and so forth :-D

and then came telemetry streaming which will hopefully be the end of snmp in general at some point.

Reply:
I think we are getting sidetracked here. Support for v3 alone could be nice. @gmuloc

Reply:
It works already and is part of the tests: https://github.com/aristanetworks/ansible-avd/pull/1721/files#diff-d65839360352a741a4aaf63924e24a87127c419b98ff93f2758a748f3aaf744cR164

Reply:
line 167 in the same file is also there to check that when localized is used but no auth or priv are given only v3 remains.

Reply:
sorry I missed that.

Reply:
@ccsnw - regarding removing the localized keywords to match the actual running-config rather than what is configurable - I just verified two extra points

1. Behavior is the same for local and remote users
2. I check what happens in CVP as if CVP were to push the localized keywords it should be seeing the device out of sync.

I Created this Configlet:

```
snmp-server user CVP-TEST CVP-GROUP v3 localized 42424242424242
```

and this is the configuration generated (I noticed that it adds the System Mac as local engine ID by itself which is not supported in this PR though I guess it _could_ be considered as a future optimization):
![Screenshot 2022-04-21 at 09 28 38](https://user-images.githubusercontent.com/3007422/164401663-768f8bd3-777b-4cc7-93b9-596f3b559035.png)

so as discussed with Claus earlier I feel confident this can be removed in this situation ( and there is a molecule test to check the behavior)

Reply:
@gmuloc 

I'd guess that EOS creates/sets the local engineID (sysmac) when you create any snmpv3 user and dont explicitly set the engineID yourself.

For remote agents, EOS cannot do that (because it doesnt know the sysmac of the remote system) and hence you have to define it manually and before remote user creation

==

anout the compliance: it's true, if we allow AVD to render "snmp-server user ... v3 localized..." and push via CVP, this will lead to a compliance error, because EOS never have a line like "snmp-server user bla test v3 321313123131" in the running-config.

My personal feeling is to put as little logic in here as possible and let the user make the choices..

Reply:
Indeed you are right, by default the engineID is derived from the System Mac:
```
Leaf2#show run | i engine
Leaf2#show snmp engineID
Local SNMP EngineID: f5717f502200715bad00
Leaf2#show version | i MAC
Hardware MAC address: 5022.0071.5bad
```

so `f5117f + < system mac > + 00`.

Regarding the CVP test, CVP is actually removing the `localized` value from the config it pushes if there is no `auth` following so in the end the only location where the localized engineID would be present is in the AVD configuration

Reply:
"Regarding the CVP test, CVP is actually removing the localized value from the config it pushes if there is no auth following so in the end the only location where the localized engineID would be present is in the AVD configuration"
>> Yup, but cvp will not modify the configlet right?
>> so if our render has "... v3 localized xxx" then cvp only pushes the "... v3" part and it will complain that designed and running config are out of sync?

Reply:
The configlet is not modified indeed but the designed configuration generated is with `... v3` only as in the screenshot above. Then CVP does not complain since the comparison is done between the designed config and the running config.

Reply:
got it, thx